code/iaas/auth-logic/src/main/java/io/cattle/platform/iaas/api/auth/dao/impl/AuthDaoImpl.java

```diff
@@ -129,70 +164,247 @@ public void updateAccount(Account account, String name, String kind, String exte
+        Map<TableField<AccountRecord, String>, String> properties = new HashMap<>();
+        if (name != null) {
+            properties.put(ACCOUNT.NAME, name);
+        }
+        if (description != null) {
+            properties.put(ACCOUNT.DESCRIPTION, description);
+        }
+        if (name == null && description == null){
+            return project;
+        }
+        int updateCount = create()
+                .update(ACCOUNT)
+                .set(properties)
+                .where(ACCOUNT.ID
+                        .eq(project.getId()))
+                .execute();
+        if(1 != updateCount) {
+            throw new RuntimeException("UpdateProject failed.");
+        }
+        return getAccountById(project.getId());
     }
 
-    public GenericResourceDao getGenericResourceDao() {
-        return resourceDao;
+    @Override
+    public List<? extends ProjectMember> setProjectMembers(final Account project, final HashSet<Member> members) {
+        return lockManager.lock(new ProjectLock(project), new LockCallback<List<? extends ProjectMember>>() {
+            public List<? extends ProjectMember> doWithLock() {
+                List<? extends ProjectMember> previousMembers = getActiveProjectMembers(project.getId());
+                HashSet<Member> otherPreviosMembers = new HashSet<>();
+                for (ProjectMember member : previousMembers) {
+                    otherPreviosMembers.add(new Member(member));
+                }
+                HashSet<Member> update = new HashSet<>();
+                HashSet<Member> createIgnore = new HashSet<>();
+                HashSet<Member> deleteIgnore = new HashSet<>();
+                for (Member member : otherPreviosMembers) {
+                    for (Member memberNew : members) {
+                        if (memberNew.getExternalId().equalsIgnoreCase(member.getExternalId()) &&
+                                memberNew.getExternalIdType().equalsIgnoreCase(member.getExternalIdType())) {
+                            if (!memberNew.getRole().equalsIgnoreCase(member.getRole())) {
+                                update.add(memberNew);
+                                deleteIgnore.add(member);
+                            } else {
+                                createIgnore.add(memberNew);
+                                deleteIgnore.add(member);
+                            }
+                        }
+                    }
+                }
+                HashSet<Member> create = (HashSet) members.clone();
+                create.removeAll(update);
+                create.removeAll(createIgnore);
+                HashSet<Member> delete = (HashSet) otherPreviosMembers.clone();
+                delete.removeAll(deleteIgnore);
+
+                Condition allMembers = DSL.falseCondition();
+                for (Member member : delete) {
+                    allMembers = allMembers.or(PROJECT_MEMBER.EXTERNAL_ID.eq(member.getExternalId())
```

Comment:
The 'delete' set can actually be a Set<ProjectMember> instead of Set<Member> since it's being populated from 'previousMembers' above.  With a Set<ProjectMember>, we can then simplify the DB query using a simple IN clause since we can use the unique IDs for each ProjectMember (rather than matching on external_id and external_id_type again).